File: decorator.java
// The Decorator design pattern is used to add new functionality to an object dynamically without altering its structure. Here's a complex example of the Decorator pattern in Java:

// Copy
// Component interface
interface Coffee {
    double cost();
    String description();
}

// ConcreteComponent class
class SimpleCoffee implements Coffee {
    @Override
    public double cost() {
        return 2.0; // Base cost of a simple coffee
    }

    @Override
    public String description() {
        return "Simple Coffee";
    }
}

// Decorator abstract class
abstract class CoffeeDecorator implements Coffee {
    protected Coffee decoratedCoffee;

    public CoffeeDecorator(Coffee decoratedCoffee) {
        this.decoratedCoffee = decoratedCoffee;
    }

    @Override
    public double cost() {
        return decoratedCoffee.cost();
    }

    @Override
    public String description() {
        return decoratedCoffee.description();
    }
}

// ConcreteDecorator classes
class MilkDecorator extends CoffeeDecorator {
    public MilkDecorator(Coffee decoratedCoffee) {
        super(decoratedCoffee);
    }

    @Override
    public double cost() {
        return super.cost() + 1.0; // Adding cost of milk
    }

    @Override
    public String description() {
        return super.description() + " with Milk";
    }
}

class SugarDecorator extends CoffeeDecorator {
    public SugarDecorator(Coffee decoratedCoffee) {
        super(decoratedCoffee);
    }

    @Override
    public double cost() {
        return super.cost() + 0.5; // Adding cost of sugar
    }

    @Override
    public String description() {
        return super.description() + " with Sugar";
    }
}

class VanillaDecorator extends CoffeeDecorator {
    public VanillaDecorator(Coffee decoratedCoffee) {
        super(decoratedCoffee);
    }

    @Override
    public double cost() {
        return super.cost() + 1.5; // Adding cost of vanilla
    }

    @Override
    public String description() {
        return super.description() + " with Vanilla";
    }
}

// Client code
public class DecoratorPatternExample {
    public static void main(String[] args) {
        Coffee simpleCoffee = new SimpleCoffee();
        System.out.println("Cost: $" + simpleCoffee.cost() + ", Description: " + simpleCoffee.description());

        Coffee milkCoffee = new MilkDecorator(simpleCoffee);
        System.out.println("Cost: $" + milkCoffee.cost() + ", Description: " + milkCoffee.description());

        Coffee sugarMilkCoffee = new SugarDecorator(milkCoffee);
        System.out.println("Cost: $" + sugarMilkCoffee.cost() + ", Description: " + sugarMilkCoffee.description());

        Coffee vanillaSugarMilkCoffee = new VanillaDecorator(sugarMilkCoffee);
        System.out.println("Cost: $" + vanillaSugarMilkCoffee.cost() + ", Description: " + vanillaSugarMilkCoffee.description());
    }
}
// In this example:

// The Coffee interface defines the component that we want to decorate.
// SimpleCoffee is a concrete component that implements the Coffee interface.
// CoffeeDecorator is an abstract class that implements the Coffee interface and holds a reference to a Coffee object.
// MilkDecorator, SugarDecorator, and VanillaDecorator are concrete decorator classes that extend CoffeeDecorator and add specific functionality (cost and description) to the decorated coffee.
// The client code demonstrates how different decorators can be combined to create a variety of coffee options. The decorators add additional behavior (cost and description) to the base SimpleCoffee object dynamically.

File: decorator.go
// Certainly! Here's an example of the Decorator design pattern implemented in Go:

// Copy
package main

import "fmt"

// Component interface
type Coffee interface {
	Cost() float64
	Description() string
}

// ConcreteComponent struct
type SimpleCoffee struct{}

func (c *SimpleCoffee) Cost() float64 {
	return 2.0 // Base cost of a simple coffee
}

func (c *SimpleCoffee) Description() string {
	return "Simple Coffee"
}

// Decorator abstract class
type CoffeeDecorator struct {
	coffee Coffee
}

func (d *CoffeeDecorator) Cost() float64 {
	return d.coffee.Cost()
}

func (d *CoffeeDecorator) Description() string {
	return d.coffee.Description()
}

// ConcreteDecorator classes
type MilkDecorator struct {
	*CoffeeDecorator
}

func NewMilkDecorator(coffee Coffee) *MilkDecorator {
	return &MilkDecorator{&CoffeeDecorator{coffee}}
}

func (d *MilkDecorator) Cost() float64 {
	return d.coffee.Cost() + 1.0 // Adding cost of milk
}

func (d *MilkDecorator) Description() string {
	return d.coffee.Description() + " with Milk"
}

type SugarDecorator struct {
	*CoffeeDecorator
}

func NewSugarDecorator(coffee Coffee) *SugarDecorator {
	return &SugarDecorator{&CoffeeDecorator{coffee}}
}

func (d *SugarDecorator) Cost() float64 {
	return d.coffee.Cost() + 0.5 // Adding cost of sugar
}

func (d *SugarDecorator) Description() string {
	return d.coffee.Description() + " with Sugar"
}

type VanillaDecorator struct {
	*CoffeeDecorator
}

func NewVanillaDecorator(coffee Coffee) *VanillaDecorator {
	return &VanillaDecorator{&CoffeeDecorator{coffee}}
}

func (d *VanillaDecorator) Cost() float64 {
	return d.coffee.Cost() + 1.5 // Adding cost of vanilla
}

func (d *VanillaDecorator) Description() string {
	return d.coffee.Description() + " with Vanilla"
}

// Client code
func main() {
	simpleCoffee := &SimpleCoffee{}
	printCoffee(simpleCoffee)

	milkCoffee := NewMilkDecorator(simpleCoffee)
	printCoffee(milkCoffee)

	sugarMilkCoffee := NewSugarDecorator(milkCoffee)
	printCoffee(sugarMilkCoffee)

	vanillaSugarMilkCoffee := NewVanillaDecorator(sugarMilkCoffee)
	printCoffee(vanillaSugarMilkCoffee)
}

func printCoffee(coffee Coffee) {
	fmt.Printf("Cost: $%.2f, Description: %s\n", coffee.Cost(), coffee.Description())
}
In this example:

Coffee is the component interface.
SimpleCoffee is the concrete component implementing the Coffee interface.
CoffeeDecorator is the decorator abstract class, embedding the Coffee interface to ensure it implements the Coffee methods.
MilkDecorator, SugarDecorator, and VanillaDecorator are concrete decorator classes that embed CoffeeDecorator and add specific functionality to the decorated coffee.
The client code demonstrates how different decorators can be combined to create various coffee options. The decorators add additional behavior (cost and description) to the base SimpleCoffee object dynamically.